### PR TITLE
Make MakeMaker pass compilation tests on AIX again

### PR DESCRIPTION
**Pull request by Leont**
We weren't taking XSMULTI into account, causing a mismatch between de dependency of dynamic/static and the actual filename.

Tux++ for giving me access to AIX

### CODE REVIEW

**Comment by mohawk2** [comment]
Looks good to me.

**Comment by jhi** [comment]
As far as I can tell... this patch changes nothing for me, sorry.  (I have not git access in this AIX box, so I may well have goofed something, do not despair.)

Is this patch based on what's on blead?

What is the right way to try to run the MM tests manually from the command line while within blead build?

I tried cd-ing to different directories, adjusting -I appropriately, using env PERL_CORE=1, all to no avail.

**Comment by Leont** [comment]
> As far as I can tell... this patch changes nothing for me, sorry. (I have not git access in this AIX box, so I may well have goofed something, do not despair.)

I tested it against the main perl on Tux's AIX machine (`/usr/opt/perl5/bin/perl`), the bug shouldn't be peculiar to blead but I'm also seeing weird issues when trying to integrate it.

> What is the right way to try to run the MM tests manually from the command line while within blead build?

I use `TEST_FILES=../cpan/ExtUtils-MakeMaker/t/02-xsdynamic.t make test`

**Comment by jhi** [comment]
Repatched, retried and still looks bad (actually might look even worse):

TEST_FILES=../cpan/ExtUtils-MakeMaker/t/02-xsdynamic.t make test
...
         ./runtests choose
t/../cpan/ExtUtils-MakeMaker/t/02-xsdynamic ... #   Failed test '"make"   exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 369.
#          got: '512'
#     expected: '0'
# cp lib/XS/Test.pm blib/lib/XS/Test.pm
# cp lib/XS/Test_BS blib/lib/XS/Test_BS
# cp lib/XS/header.h blib/lib/XS/header.h
# cp lib/XS/typemap blib/lib/XS/typemap
# cp lib/XS/Other.pm blib/lib/XS/Other.pm
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Other]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Test]);"
# Running Mkbootstrap for lib/XS/Other ()
#       chmod 644 "lib/XS/Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Other.bs blib/arch/auto/XS/Other/Other.bs 644
# Running Mkbootstrap for lib/XS/Test ()
# Writing lib/XS/Test.bs
#       chmod 644 "lib/XS/Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testbscodemulti/../../../../lib/ExtUtils/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testbscodemulti/lib/XS/typemap'  lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testbscodemulti/../../../../lib/ExtUtils/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testbscodemulti/lib/XS/typemap' lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../.."   lib/XS/Other.c -o lib/XS/Other.o
#       rm -f blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib lib/XS/Other.o  -o blib/arch/auto/XS/Other/Other.so        
# ld: 0706-004 Cannot find or read export file: Test.exp
#       ld:accessx(): A file or directory in the path name does not exist.
# make: 1254-004 The error code from the last command is 255.
# 
# 
# Stop.
#   Failed test '"make" test  exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 390.
#          got: '512'
#     expected: '0'
# cp Test.pm blib/lib/XS/Test.pm
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
# Running Mkbootstrap for Test ()
#       chmod 644 "Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testsubdirscomplex/../../../../lib/ExtUtils/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testsubdirscomplex/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testsubdirscomplex/typemap'  Test.xs > Test.xsc
#       mv Test.xsc Test.c
#       cc -q32 -c  -Inewline  -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.01\"  -DXS_VERSION=\"1.01\"  "-I../../../.."  -DINVAR=input Test.c
#       rm -f blib/arch/auto/XS/Test/Test.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib Test.o  -o blib/arch/auto/XS/Test/Test.so        
#       chmod 755 blib/arch/auto/XS/Test/Test.so
#       cd lib && cc -q32 -c -Inewline    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../../.." -DINVAR=input   ../lib/file.c
# 
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
# Running Mkbootstrap for Other ()
#       chmod 644 "Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- Other.bs ../blib/arch/auto/XS/Other/Other.bs 644
#       rm -f ../blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../../perl.exp -bE:Other.exp -bnoentry -lc -lm  -L/usr/local/lib lib/file.o  -o ../blib/arch/auto/XS/Other/Other.so        
# ld: 0711-244 ERROR: No csects or exported symbols have been saved.
# make: 1254-004 The error code from the last command is 8.
# 
# 
# Stop.
# make: 1254-004 The error code from the last command is 2.
# 
# 
# Stop.
#   Failed test '"make"   exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 369.
#          got: '512'
#     expected: '0'
# cp lib/XS/Test.pm blib/lib/XS/Test.pm
# cp lib/XS/typemap blib/lib/XS/typemap
# cp lib/XS/header.h blib/lib/XS/header.h
# cp lib/XS/Other.pm blib/lib/XS/Other.pm
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Other]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Test]);"
# Running Mkbootstrap for lib/XS/Other ()
#       chmod 644 "lib/XS/Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Other.bs blib/arch/auto/XS/Other/Other.bs 644
# Running Mkbootstrap for lib/XS/Test ()
#       chmod 644 "lib/XS/Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testmulti/../../../../lib/ExtUtils/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testmulti/lib/XS/typemap'  lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testmulti/../../../../lib/ExtUtils/typemap' -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testmulti/lib/XS/typemap' lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../.."   lib/XS/Other.c -o lib/XS/Other.o
#       rm -f blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib lib/XS/Other.o  -o blib/arch/auto/XS/Other/Other.so        
# ld: 0706-004 Cannot find or read export file: Test.exp
#       ld:accessx(): A file or directory in the path name does not exist.
# make: 1254-004 The error code from the last command is 255.
# 
# 
# Stop.
#   Failed test '"make" test LINKTYPE=dynamic exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 390.
#          got: '512'
#     expected: '0'
# cp lib/XS/Other.pm blib/lib/XS/Other.pm
# cp lib/XS/header.h blib/lib/XS/header.h
# cp lib/XS/typemap blib/lib/XS/typemap
# cp lib/XS/Test.pm blib/lib/XS/Test.pm
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Other]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Test]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-TeststaticmultiLINKTYPEdynamic/../../../../lib/ExtUtils/typemap'  lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-TeststaticmultiLINKTYPEdynamic/../../../../lib/ExtUtils/typemap' lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../.."   lib/XS/Other.c -o lib/XS/Other.o
#       rm -f "blib/arch/auto/XS/Other/Other.a"
#       /usr/bin/ar cr "blib/arch/auto/XS/Other/Other.a" lib/XS/Other.o
#       : "blib/arch/auto/XS/Other/Other.a"
#       chmod 755 blib/arch/auto/XS/Other/Other.a
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-TeststaticmultiLINKTYPEdynamic/../../../../lib/ExtUtils/typemap'  lib/XS/Test.xs > lib/XS/Test.xsc
#       mv lib/XS/Test.xsc lib/XS/Test.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-TeststaticmultiLINKTYPEdynamic/../../../../lib/ExtUtils/typemap' lib/XS/Test.xs > lib/XS/Test.xsc
#       mv lib/XS/Test.xsc lib/XS/Test.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.01\"  -DXS_VERSION=\"1.01\"  "-I../../../.."   lib/XS/Test.c -o lib/XS/Test.o
#       rm -f "blib/arch/auto/XS/Test/Test.a"
#       /usr/bin/ar cr "blib/arch/auto/XS/Test/Test.a" lib/XS/Test.o
#       : "blib/arch/auto/XS/Test/Test.a"
#       chmod 755 blib/arch/auto/XS/Test/Test.a
# 
# Running Mkbootstrap for lib/XS/Other ()
#       chmod 644 "lib/XS/Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Other.bs blib/arch/auto/XS/Other/Other.bs 644
# Running Mkbootstrap for lib/XS/Test ()
#       chmod 644 "lib/XS/Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       rm -f blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib lib/XS/Other.o  -o blib/arch/auto/XS/Other/Other.so        
# ld: 0706-004 Cannot find or read export file: Test.exp
#       ld:accessx(): A file or directory in the path name does not exist.
# make: 1254-004 The error code from the last command is 255.
# 
# 
# Stop.
#   Failed test '"make"  dynamic  exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 369.
#          got: '512'
#     expected: '0'
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Other]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Test]);"
# Running Mkbootstrap for lib/XS/Other ()
# Target "dynamic" is up to date.
#       chmod 644 "lib/XS/Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Other.bs blib/arch/auto/XS/Other/Other.bs 644
# Running Mkbootstrap for lib/XS/Test ()
#       chmod 644 "lib/XS/Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Teststaticmultidynamic/../../../../lib/ExtUtils/typemap'  lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Teststaticmultidynamic/../../../../lib/ExtUtils/typemap' lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../.."   lib/XS/Other.c -o lib/XS/Other.o
#       rm -f blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib lib/XS/Other.o  -o blib/arch/auto/XS/Other/Other.so        
# ld: 0706-004 Cannot find or read export file: Test.exp
#       ld:accessx(): A file or directory in the path name does not exist.
# make: 1254-004 The error code from the last command is 255.
# 
# 
# Stop.
#   Failed test '"make"   exited normally'
#   at lib/MakeMaker/Test/Setup/XS.pm line 369.
#          got: '512'
#     expected: '0'
# cp lib/XS/header.h blib/lib/XS/header.h
# cp lib/XS/Other.pm blib/lib/XS/Other.pm
# cp lib/XS/Test.pm blib/lib/XS/Test.pm
# cp lib/XS/typemap blib/lib/XS/typemap
# cp lib/XS/plus1.c blib/lib/XS/plus1.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Other\", 'DLBASE' => 'Other', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Other]);"
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Mksymlists  -e "Mksymlists('NAME'=>\"XS::Test\", 'DLBASE' => 'Test', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => [], 'FILE' => q[lib/XS/Test]);"
# Running Mkbootstrap for lib/XS/Other ()
#       chmod 644 "lib/XS/Other.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Other.bs blib/arch/auto/XS/Other/Other.bs 644
# Running Mkbootstrap for lib/XS/Test ()
#       chmod 644 "lib/XS/Test.bs"
#        "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" -MExtUtils::Command::MM -e 'cp_nonempty' -- lib/XS/Test.bs blib/arch/auto/XS/Test/Test.bs 644
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testxsbuild/../../../../lib/ExtUtils/typemap'  lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       "/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/../../t/perl" -w "-I../../../../lib" "../../../../lib/ExtUtils/xsubpp"  -typemap '/perl/usr/jhi/perl/cpan/ExtUtils-MakeMaker/t/XS-Testxsbuild/../../../../lib/ExtUtils/typemap' lib/XS/Other.xs > lib/XS/Other.xsc
#       mv lib/XS/Other.xsc lib/XS/Other.c
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.2\"  -DXS_VERSION=\"1.2\"  "-I../../../.."  -DINVAR=input lib/XS/Other.c -o lib/XS/Other.o
#       cc -q32 -c    -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -qmaxmem=-1 -qnoansialias -qlanglvl=extc99 -DUSE_NATIVE_DLOPEN -I/usr/local/include -q32 -D_LARGE_FILES -O    -DVERSION=\"1.01\"  -DXS_VERSION=\"1.01\"  "-I../../../.."   lib/XS/plus1.c -o lib/XS/plus1.o
#       rm -f blib/arch/auto/XS/Other/Other.so
#       ld  -bhalt:4 -G -bI:../../../../perl.exp -bE:Test.exp -bnoentry -lc -lm  -L/usr/local/lib lib/XS/Other.o lib/XS/plus1.o  -o blib/arch/auto/XS/Other/Other.so        
# ld: 0706-004 Cannot find or read export file: Test.exp
#       ld:accessx(): A file or directory in the path name does not exist.
# make: 1254-004 The error code from the last command is 255.
# 
# 
# Stop.
# Looks like you failed 6 tests of 66.
FAILED at test 16
Failed 1 test out of 1, 0.00% okay.
        ../cpan/ExtUtils-MakeMaker/t/02-xsdynamic.t
### Since not all tests were successful, you may want to run some of
### them individually and examine any diagnostic messages they produce.
### See the INSTALL document's section on "make test".
### You may have to set your dynamic library search path,
### LIBPATH, to point to the build directory:
###   setenv LIBPATH `pwd`; cd t; ./perl harness
###   LIBPATH=`pwd`; export LIBPATH; cd t; ./perl harness
###   export LIBPATH=`pwd`; cd t; ./perl harness
### for csh-style shells, like tcsh; or for traditional/modern
### Bourne-style shells, like bash, ksh, and zsh, respectively.
Elapsed: 44 sec
u=0.01  s=0.00  cu=18.21  cs=3.82  scripts=1  tests=66
make: 1254-004 The error code from the last command is 1.

**Comment by Leont** [comment]
> Repatched, retried and still looks bad

**Comment by Leont** [comment]
I don't understand why I'm getting such a difference on the dist but not in core, as far as I can see now it shouldn't work on the dist either.

The problem is lddlflags, it's value on AIX is typically `-bhalt:4 -G -bI:$(PERL_INC)/perl.exp -bE:$(BASEEXT).exp -bnoentry -lpthreads -lc -lm`. Under XSMULTI it would need to be adapted to each xs target.

**Comment by mohawk2** [comment]
With Leon's agreement, I've taken over his branch and therefore this PR with a slightly different approach, which I'll document after this point:

There was a problem with "dlsyms" file-generation on AIX even prior to `XSMULTI`: the `*.exp` file was a direct dependency of `dynamic`, but used in generating a DLL, which was also a direct dep from `dynamic`. The "correct" dependency should be that the DLL depends on the `*.exp` file(s). This meant that parallel builds would be unreliable. Leon said he had occasionally seen this.

Leon's initial approach to this problem made the dlsyms files all be generated, but still as a dep from `dynamic`. This improved matters but does not solve the dependency/parallel-build issue identified above. Instead, the approach I am taking is to make each `*.exp` file depend on its relevant DLL.

This leads to only 6/11 fails on AIX. The remaining failures are, remarkably, of the non-`XSMULTI` cases. It turns out that the DLL-linking command is getting no `*.o` files to link. The error from the linker did not actually say this however!

**Comment by mohawk2** [comment]
That last part was due to a full override of `init_others` without calling `SUPER` first. Also, `subdirscomplex` needs disabling like on VMS.

Now ready for review.

File: lib/ExtUtils/MM_AIX.pm [hunk 1]
```diff
@@ -53,6 +48,21 @@ sub xs_dlsyms_ext {
     '.exp';
 }
 
+sub xs_dlsyms_arg {
+    my($self, $file) = @_;
+    return qq{-bE:${file}};
+}
+
+sub init_others {
+    my $self = shift;
```
Comment (mohawk2) [inline lib/ExtUtils/MM_AIX.pm]: Belatedly, I agree. I copy/pasted it from another `MM_*` which had that.

File: lib/ExtUtils/MM_AIX.pm [hunk 2]
```diff
@@ -53,6 +48,21 @@ sub xs_dlsyms_ext {
...
+    $self->SUPER::init_others;
+    # perl "hints" add -bE:$(BASEEXT).exp to LDDLFLAGS. strip that out
+    # so right value can be added by xs_make_dynamic_lib to work for XSMULTI
+    $self->{LDDLFLAGS} ||= $Config{lddlflags};
+    $self->{LDDLFLAGS} =~ s#(\s*)\S*\Q$(BASEEXT)\E\S*(\s*)#$1$2#;
```
Comment (mohawk2) [inline lib/ExtUtils/MM_AIX.pm]: I felt that with only three "parts" to it, two of which were just `\s*`, that the previous comment was sufficient explanation. We can always make another PR if this is truly unclear.

File: lib/ExtUtils/MM_AIX.pm [hunk 3]
```diff
@@ -53,6 +48,21 @@ sub xs_dlsyms_ext {
     '.exp';
 }
 
+sub xs_dlsyms_arg {
+    my($self, $file) = @_;
+    return qq{-bE:${file}};
```
Comment (mohawk2) [inline lib/ExtUtils/MM_AIX.pm]: This was previously a slightly longer thing, and I kept the highly defensive `${}`.

File: lib/ExtUtils/MM_AIX.pm (hunk 2 above)
Comment (wchristian) [inline lib/ExtUtils/MM_AIX.pm]: I'd like to see one for that. This is not overly complex, but the code is sufficiently unclear that some unpacking of what it does is merited, if only to save future readers the time to unpack.

**Comment by kentfredric** [comment]
This PR might be to blame for https://rt.cpan.org/Ticket/Display.html?id=122554 , but its not completely clear.

**Comment by Leont** [comment]
>  bingos merged commit 982f991 into master on May 15, 2017 

I guess that was a squash-and-merge, because it now attributes code to me that I'm sure I didn't write!
